[RF-DOCS] Action View Form Helpers Guide [ci-skip] (#51936)
rails/rails: guides/source/form_helpers.md

The first parameter to these methods is always the name of the input. This is
useful to remember because when the form is submitted, that name will be passed
to the controller along with the form data in the `params` hash. The name will be the key in the `params` for the value entered by the user for that field.

For example, if the form contains `<%= form.text_field :query %>`, then you
would be able to get the value of this field in the controller with
`params[:query]`.

When naming inputs, Rails uses certain conventions that make it possible to submit parameters with non-scalar values such as arrays or hashes, which will also be accessible in `params`. You can read more about them in the [Form Input Naming Conventions and Params Hash](#form-input-naming-conventions-and-params-hash) section of this guide. For details on the precise usage of these helpers, please refer to the [API documentation](https://api.rubyonrails.org/classes/ActionView/Helpers/FormTagHelper.html).

#### Checkboxes

A Checkbox is a form control that allows for a single value to be selected or deselected. A group of Checkboxes is generally used to allow a user to choose one or more options from the group.

> For example, two checkboxes in a form:

Reviewer: Ins't three per the example below?

<label for="biography">Biography</label>
<input name="romance" type="hidden" value="0" autocomplete="off"><input type="checkbox" value="1" name="romance" id="romance">
<label for="romance">Romance</label>
<input name="mystery" type="hidden" value="0" autocomplete="off"><input type="checkbox" value="1" name="mystery" id="mystery">
<label for="mystery">Mystery</label>
```

The first parameter to [`check_box`](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-check_box) is the name of the input which can be found in the `params` hash. If the user has checked the "Biography" checkbox only, the `params` hash would contain `params: {"biography"=>"1", "romance"=>"0", "mystery"=>"0"}`. You can use `params[:biography]` to check if that checkbox is selected by the user.

The checkbox's values (the values that will appear in `params`) can optionally be specified using the `checked_value` and `unchecked_value` parameters. See the [API documentation](https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-check_box) for more details.

> 

Reviewer 2: I know that this is a basics guide and not intended to be comprehensive, but it might be worth briefly mentioning `collection_check_boxes` here, as implementing multiple check-boxes as part of a collection  of items is something that users often struggle with. Even without an example, just highlighting that this method exists might be helpful.

Reviewer 3: `collection_check_boxes` are mentioned later on in the guide.
We could link to that section.

Author: linked

<%= form.label :flavor_vanilla, "Vanilla" %>
<%= form.radio_button :flavor, "hazelnut" %>
<%= form.label :flavor_hazelnut, "Hazelnut" %>
```

The above will generate the following HTML:

```html
<input type="radio" value="chocolate_chip" name="flavor" id="flavor_chocolate_chip">
<label for="flavor_chocolate_chip">Chocolate Chip</label>
<input type="radio" value="vanilla" name="flavor" id="flavor_vanilla">
<label for="flavor_vanilla">Vanilla</label>
<input type="radio" value="hazelnut" name="flavor" id="flavor_hazelnut">
<label for="flavor_hazelnut">Hazelnut</label>
```

The second parameter to [`radio_button`](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-radio_button) is the value of the input. Because these radio buttons share the same name (`flavor`), the user will only be able to select one of them, and `params[:flavor]` will contain either `"chocolate_chip"`, `"vanilla"`, or `hazelnut`.

NOTE: Always use labels for checkbox and radio buttons. They associate text with
a specific option using the `for` attribute and, by expanding the clickable
region, make it easier for users to click the inputs.

### Other Helpers of Interest

> There are many other form controls for text, email, password, date, and time. Below examples show some more helpers and their generated HTML.

Reviewer 2: The way this sentence is worded perhaps suggests that this list is exhaustive (which it isn't). Maybe something like this would be clearer:

```suggestion
There are many other form controls including text, email, password, date, and time. The below examples show some more helpers and their generated HTML.
```

<%= form.month_field :birthday_month %>
<%= form.week_field :birthday_week %>
```

Output:

```html
<input type="date" name="born_on" id="born_on">
<input type="time" name="started_at" id="started_at">
<input type="datetime-local" name="graduation_day" id="graduation_day">
<input type="month" name="birthday_month" id="birthday_month">
<input type="week" name="birthday_week" id="birthday_week">
```

Helpers with special formatting:

```erb
<%= form.password_field :password %>
<%= form.email_field :address %>
<%= form.telephone_field :phone %>
<%= form.url_field :homepage %>
```

Output:

```html
<input type="password" name="password" id="password">
<input type="email" name="address" id="address">
<input type="tel" name="phone" id="phone">
<input type="url" name="homepage" id="homepage">
```

Other common helpers:

```erb
<%= form.text_area :message, size: "70x5" %>
<%= form.hidden_field :parent_id, value: "foo" %>
<%= form.number_field :price, in: 1.0..20.0, step: 0.5 %>
<%= form.range_field :discount, in: 1..100 %>
<%= form.search_field :name %>
<%= form.color_field :favorite_color %>
```

Output:

```html
<textarea name="message" id="message" cols="70" rows="5"></textarea>

> <input value="foo" autocomplete="off" type="hidden" name="parent_id" id="parent_id">

Reviewer: It would be great to set `type` as first attribute to be consistent with other examples.

Reviewer 3: I agree it would look nice, but this currently is the actual output of these helpers.
Maybe we should change the implementation to render these consistently instead.

Author: Yeah, I copy/pasted output from an app.

<input step="0.5" min="1.0" max="20.0" type="number" name="price" id="price">
<input min="1" max="100" type="range" name="discount" id="discount">
<input type="search" name="name" id="name">
<input value="#000000" type="color" name="favorite_color" id="favorite_color">
```

Hidden inputs are not shown to the user but instead hold data like any textual input. Values inside them can be changed with JavaScript.

TIP: If you're using password input fields, you might want to configure your application to prevent those parameters from being logged. You can learn about how in the [Securing Rails Applications](security.html#logging) guide.

Creating Forms with Model Objects
---------------------------------

### Binding a Form to an Object

The `form_with` helper has a `:model` option that allows you to bind the form builder object to a model object. This means that the form will be scoped to that model object, and the form's fields will be populated with values from that model object.

> For example, if we have an `@book` model object:

Reviewer: _if we have a `@book`_ perhaps?

```ruby
@book = Book.find(42)
# => #<Book id: 42, title: "Walden", author: "Henry David Thoreau">
```

And the following form to create a new book:

```erb
<%= form_with model: @book do |form| %>
  <div>
    <%= form.label :title %>
    <%= form.text_field :title %>
  </div>
  <div>
    <%= form.label :author %>
    <%= form.text_field :author %>
  </div>
  <%= form.submit %>
<% end %>
```

It will generate this HTML:

```html
<form action="/books" accept-charset="UTF-8" method="post">
  <input type="hidden" name="authenticity_token" value="ChwHeyegcpAFDdBvXvDuvbfW7yCA3e8gvhyieai7DhG28C3akh-dyuv-IBittsjPrIjETlQQvQJ91T77QQ8xWA" autocomplete="off">
  <div>
    <label for="book_title">Title</label>
    <input type="text" name="book[title]" id="book_title">
  </div>
  <div>
    <label for="book_author">Author</label>
    <input type="text" name="book[author]" id="book_author">
  </div>
  <input type="submit" name="commit" value="Create Book" data-disable-with="Create Book">
</form>
```

> Some important things to note when using `form_with` with a model object:

Reviewer: _Some important things to notice_

#### Checkboxes

A Checkbox is a form control that allows for a single value to be selected or deselected. A group of Checkboxes is generally used to allow a user to choose one or more options from the group.

Here's an example with three checkboxes in a form:

```erb
<%= form.check_box :biography %>
<%= form.label :biography, "Biography" %>
<%= form.check_box :romance %>
<%= form.label :romance, "Romance" %>
<%= form.check_box :mystery %>
<%= form.label :mystery, "Mystery" %>
```

The above will generate the following:

```html
<input name="biography" type="hidden" value="0" autocomplete="off"><input type="checkbox" value="1" name="biography" id="biography">
<label for="biography">Biography</label>
<input name="romance" type="hidden" value="0" autocomplete="off"><input type="checkbox" value="1" name="romance" id="romance">
<label for="romance">Romance</label>
<input name="mystery" type="hidden" value="0" autocomplete="off"><input type="checkbox" value="1" name="mystery" id="mystery">
<label for="mystery">Mystery</label>
```

> The first parameter to [`check_box`](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-check_box) is the name of the input which can be found in the `params` hash. If the user has checked the "Biography" checkbox only, the `params` hash would contain `params: {"biography"=>"1", "romance"=>"0", "mystery"=>"0"}`. You can use `params[:biography]` to check if that checkbox is selected by the user.

Author: Updated (manually, so I can add new line before the code fence)

<% end %>
```

> The above will produces the following output:

Reviewer: _will produce_

Reviewer 3: Hi @ruyrocha . Not sure if you are familiar with the suggestion button?
This would allow you add suggestion than can merged with a single click.
<img width="800" alt="image" src="https://github.com/rails/rails/assets/28561/5ff0b23a-f3ab-4b46-9f3f-43de0416ca6b">

```html
<form action="/people" accept-charset="UTF-8" method="post">

>   <input type="hidden" name="authenticity_token" value="..." autocomplete="off" />

Reviewer: This isn't a big deal, but the other examples are including the non-redacted authenticity token.

Author: That's true. I added it back for the sake of consistency.

WARNING: When you're using [single-table inheritance](association_basics.html#single-table-inheritance-sti) with your models, you can't rely on record identification on a subclass if only their parent class is declared a resource. You will have to specify `:url`, and `:scope` (the model name) explicitly.

### Working with Namespaces

If you have namespaced routes, `form_with` has a shorthand for that. For example, if your application has an `admin` namespace:

```ruby
form_with model: [:admin, @article]
```

The above will create a form that submits to the `Admin::ArticlesController` inside the admin namespace,  therefore submitting to `admin_article_path(@article)` in the case of an update.

If you have several levels of namespacing then the syntax is similar:

```ruby
form_with model: [:admin, :management, @article]
```

For more information on Rails' routing system and the associated conventions, please see the [Rails Routing from the Outside In](routing.html) guide.

### Forms with PATCH, PUT, or DELETE Methods

> The Rails framework encourages RESTful design, which means forms in your application will make requests where the `method` is `PATCH`, `PUT`, or `DELETE` in addition to `GET` and `POST`. However, most browsers _don't support_ methods other than `GET` and `POST` when it comes to submitting forms.

Reviewer 3: I think this is [correct](https://github.com/whatwg/html/issues/3577).
Maybe something like:
```suggestion
The Rails framework encourages RESTful design, which means forms in your application will make requests where the `method` is `PATCH`, `PUT`, or `DELETE` in addition to `GET` and `POST`. However, HTML forms _don't support_ methods other than `GET` and `POST` when it comes to submitting forms.
```
We can probably skip the [`dialog`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#method)action for simplicity.

Uploading Files
---------------

> A common task with forms is allowing users to upload a file. It could be an avatar image or a CSV file with data to process. File upload fields can be rendered with the [`file_field`](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-file_field) helper.

Reviewer 2: Might be worth briefly mentioning the `:multiple` option for `file_field`.

Reviewer 3: `:multiple` and other options are described in the linked API docs as well.
Is it that important?

Reviewer 2: Probably not super important, tbh.

Uploading Files
---------------

A common task with forms is allowing users to upload a file. It could be an avatar image or a CSV file with data to process. File upload fields can be rendered with the [`file_field`](https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-file_field) helper.

```erb
<%= form_with model: @person do |form| %>
  <%= form.file_field :csv_file %>
<% end %>
```

> The most important thing to remember with file uploads is that the rendered form's `enctype` attribute **must** be set to "multipart/form-data". This is done automatically if you use a `file_field` inside a `form_with`. You can also set the attribute manually:

Reviewer 3: For consistency maybe use backticks here as well?
```suggestion
The most important thing to remember with file uploads is that the rendered form's `enctype` attribute **must** be set to `multipart/form-data`. This is done automatically if you use a `file_field` inside a `form_with`. You can also set the attribute manually:
```

<input id="person_address_city" name="person[address][city]" type="text" value="New York"/>
```

The above will result in the `params` hash being

```ruby
{ 'person' => { 'address' => { 'city' => 'New York' } } }
```

The other structure is an Array. Normally Rails ignores duplicate parameter names, but if the parameter name ends with an empty set of square brackets `[]` then the parameters will be accumulated in an Array.

For example, if you want users to be able to input multiple phone numbers, you could place this in the form:

```html
<input name="person[phone_number][]" type="text"/>
<input name="person[phone_number][]" type="text"/>
<input name="person[phone_number][]" type="text"/>
```

> This would result in `params[:person][:phone_number]` being an array containing the submitted phone numbers.

Reviewer 3: For consistency it would be nice to show the result here as well:
````suggestion
This would result in `params[:person][:phone_number]` being an array containing the submitted phone numbers.

```ruby
{ 'person' => { 'phone_number' => ['555-0123', '555-0124', '555-0125'] } }
```
````
```

Author: Updated.

<input name="person[addresses][][city]" type="text"/>
```

This would result in `params[:person][:addresses]` being an array of hashes. Each hash in the array will have the keys `line1`, `line2`, and `city`.

It's important to note that while hashes can be nested arbitrarily, only one level of "arrayness" is allowed. Arrays can usually be replaced by hashes. For example, instead of an array of model objects, you can have a hash of model objects keyed by their id or similar.

> WARNING: Array parameters do not play well with the `check_box` helper. According to the HTML specification unchecked checkboxes submit no value. However it is often convenient for a checkbox to always submit a value. The `check_box` helper fakes this by creating an auxiliary hidden input with the same name. If the checkbox is unchecked only the hidden input is submitted. If it is checked then both are submitted but the value submitted by the checkbox takes precedence.

Reviewer 2: Perhaps mention the `:include_hidden` option here?

Author: added

<input name="person[addresses][][city]" type="text"/>
```

> This would result in `params[:person][:addresses]` being an array of hashes. Each hash in the array will have the keys `line1`, `line2`, and `city`.

Reviewer 3: It would be nice to show the results here as well.
I haven't checked if the output is correct here...
````suggestion
This would result in `params[:person][:addresses]` being an array of hashes. Each hash in the array will have the keys `line1`, `line2`, and `city`.

```ruby
{ 'person' =>
  { 'addresses' => [
     { 'line1' => '1000 Fifth Avenue',
       'line2' => '',
       'city' => 'New York'
     },
    { 'line1' => 'Calle de Ruiz de Alarcón',
       'line2' => '',
       'city' => 'Madrid'
     }
     ]
  }
}
```
````

Author: Updated

It's important to note that while hashes can be nested arbitrarily, only one level of "arrayness" is allowed. Arrays can usually be replaced by hashes. For example, instead of an array of model objects, you can have a hash of model objects keyed by their id or similar.

> WARNING: Array parameters do not play well with the `check_box` helper. According to the HTML specification unchecked checkboxes submit no value. However it is often convenient for a checkbox to always submit a value. The `check_box` helper fakes this by creating an auxiliary hidden input with the same name. If the checkbox is unchecked only the hidden input is submitted. If it is checked then both are submitted but the value submitted by the checkbox takes precedence. There is a `include_hidden` option can be set to `false` if you want to omit this hidden field. By default, this option is `true`.

Reviewer 3: ```suggestion
WARNING: Array parameters do not play well with the `check_box` helper. According to the HTML specification unchecked checkboxes submit no value. However it is often convenient for a checkbox to always submit a value. The `check_box` helper fakes this by creating an auxiliary hidden input with the same name. If the checkbox is unchecked only the hidden input is submitted. If it is checked then both are submitted but the value submitted by the checkbox takes precedence. There is a `include_hidden` option that can be set to `false` if you want to omit this hidden field. By default, this option is `true`.
```

```ruby
{ 'person' =>
  { 'addresses' => [
    { 'line1' => '1000 Fifth Avenue',
      'line2' => '',
      'city' => 'New York'
    },
    { 'line1' => 'Calle de Ruiz de Alarcón',
      'line2' => '',
      'city' => 'Madrid'
    }
    ]
  }
}
```

It's important to note that while hashes can be nested arbitrarily, only one level of "arrayness" is allowed. Arrays can usually be replaced by hashes. For example, instead of an array of model objects, you can have a hash of model objects keyed by their id or similar.

WARNING: Array parameters do not play well with the `check_box` helper. According to the HTML specification unchecked checkboxes submit no value. However it is often convenient for a checkbox to always submit a value. The `check_box` helper fakes this by creating an auxiliary hidden input with the same name. If the checkbox is unchecked only the hidden input is submitted. If it is checked then both are submitted but the value submitted by the checkbox takes precedence. There is a `include_hidden` option that can be set to `false` if you want to omit this hidden field. By default, this option is `true`.

### The `fields_for` Helper `:index` Option

> Let's say you want to render a form with a set of fields for each of a person's
> addresses. The [`fields_for`][] helper with its `:index` option can assist:

Reviewer 3: This fields_for currently isn't linked:
<img width="821" alt="image" src="https://github.com/rails/rails/assets/28561/a4fa809f-d1e3-47f5-bf08-235b6df79bdf">

```suggestion
Let's say you want to render a form with a set of fields for each of a person's
addresses. The [`fields_for`][] helper with its `:index` option can assist:
[`fields_for`]: https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-fields_for
```

All of the form inputs map to the `"person"` hash because we called `fields_for`
on the `person_form` form builder. Also, by specifying `index: address.id`, we
rendered the `name` attribute of each city input as

> `person[address][#{address.id}][city]` instead of `person[address][city]`. This

Reviewer 3: I think this whole section should use `addresses` instead of `address` for the form/parameters.
The association on person would probably be `addresses` as well and other examples above use `addresses` as well (this suggestion doesn't include all occurences as that was a bit diffuclt to do in Github 😞 )
```suggestion
`person[addresses][#{address.id}][city]` instead of `person[addresses][city]`. This
```

Hmm, the next section implements this very similar example with nested attributes. Maybe this section should use a single address for a person instead?

Author: Hi @p8 - let me know if I misunderstood something but I double checked and it is indeed `address` as it is coming from within the loop @person.addresses.each do |address|:

```ruby
  <% @person.addresses.each do |address| %>
    <%= person_form.fields_for address, index: address.id do |address_form| %>
      <%= address_form.text_field :city %>
    <% end %>
  <% end %>
```
The `id` specified using the `index:` option is what will then distinguish the addresses with `person[address][23][city]`, I think that's the "feature" this section is trying to illustrate. Let me know if that make sense, or if I've misunderstood what you're getting it happy to look into it more.

end
```

If the hash of attributes for an object contains the key `_destroy` with a value

> that evaluates to `true` (e.g. 1, '1', true, or 'true') then the object will be

Reviewer 3: This section is probably easier to read if we use backticks as well?
```suggestion
that evaluates to `true` (e.g. `1`, `'1'`, `true`, or `'true'`) then the object will be
```
